panic during module installation with invalid provider name

Hi there,

I have a crash issue with HashiCorp Vault

**Source Example:**

### Terraform Version Terraform v1.0.11
Terraform v1.0.11
on windows_amd64

### Terraform Configuration Files

main.tf

// #########################################################
// PROVIDERS
// #########################################################
provider "azurerm" {
  subscription_id = var.subscription_id
  client_id       = var.client_id
  client_secret   = var.client_secret
  tenant_id       = var.tenant_id
  features {}
}

data "azurerm_client_config" "current" {}

./modules/hc_vault_access/main.tf

// #########################################################
// CONNECT TO VAULT
// #########################################################
module "hc_vault_access" {

  source = "./modules/hc_vault_access"

  hcv_address      = var.hcv_address
  hcv_access_token = var.hcv_access_token

}


submodule

modules/hc_vault_access

// connect to vault using root token
provider "hc_vault_access" {
  address            = var.hcv_address
  add_address_to_env = true
  token              = var.hcv_access_token
  skip_tls_verify    = true
}

### Steps to Reproduce

terraform init


### Debug Output
**Log file**
2021-12-03T13:27:54.908+0100 [INFO]  Terraform version: 1.0.11
2021-12-03T13:27:54.909+0100 [INFO]  Go runtime version: go1.16.4
2021-12-03T13:27:54.909+0100 [INFO]  CLI args: []string{"D:\\HashiCorpTerraform\\terraform.exe", "init"}
2021-12-03T13:27:54.910+0100 [TRACE] Stdout is a terminal of width 165
2021-12-03T13:27:54.910+0100 [TRACE] Stderr is a terminal of width 165
2021-12-03T13:27:54.910+0100 [TRACE] Stdin is a terminal
2021-12-03T13:27:54.914+0100 [DEBUG] Attempting to open CLI config file: C:\Users\tkp\AppData\Roaming\terraform.rc
2021-12-03T13:27:54.914+0100 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2021-12-03T13:27:54.914+0100 [INFO]  Loading CLI configuration from C:\Users\tkp\AppData\Roaming\terraform.d\credentials.tfrc.json
2021-12-03T13:27:54.915+0100 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2021-12-03T13:27:54.915+0100 [DEBUG] ignoring non-existing provider search directory C:\Users\tkp\AppData\Roaming\terraform.d\plugins
2021-12-03T13:27:54.917+0100 [DEBUG] ignoring non-existing provider search directory C:\Users\tkp\AppData\Roaming\HashiCorp\Terraform\plugins
2021-12-03T13:27:54.917+0100 [INFO]  CLI command args: []string{"init"}
2021-12-03T13:27:54.920+0100 [TRACE] ModuleInstaller: installing child modules for . into .terraform\modules
2021-12-03T13:27:54.922+0100 [DEBUG] Module installer: begin hc_vault_access
2021-12-03T13:27:54.922+0100 [TRACE] ModuleInstaller: Module installer: hc_vault_access <nil> already installed in modules\hc_vault_access
2021-12-03T13:27:54.922+0100 [TRACE] modsdir: writing modules manifest to .terraform\modules\modules.json

panic: must contain only letters, digits, and dashes, and may not use leading or trailing dashes

goroutine 1 [running]:
github.com/hashicorp/terraform/internal/addrs.MustParseProviderPart(0xc0003bf400, 0xf, 0xc0003bf400, 0xf)
	/home/circleci/project/project/internal/addrs/provider.go:449 +0xae
github.com/hashicorp/terraform/internal/addrs.NewDefaultProvider(...)
	/home/circleci/project/project/internal/addrs/provider.go:122
github.com/hashicorp/terraform/internal/configs.(*Config).resolveProviderTypes(0xc0002ce410)
	/home/circleci/project/project/internal/configs/config.go:352 +0x42e
github.com/hashicorp/terraform/internal/configs.(*Config).resolveProviderTypes(0xc000285930)
	/home/circleci/project/project/internal/configs/config.go:337 +0xab
github.com/hashicorp/terraform/internal/configs.BuildConfig(0xc00010a5a0, 0x29ddf80, 0xc000145a10, 0xc00010a500, 0x0, 0x0, 0x0)
	/home/circleci/project/project/internal/configs/config_build.go:28 +0xd1
github.com/hashicorp/terraform/internal/configs/configload.(*Loader).LoadConfig(0xc00089c7d0, 0x259572d, 0x1, 0x0, 0x1, 0x1, 0x0)
	/home/circleci/project/project/internal/configs/configload/loader_load.go:33 +0x185
github.com/hashicorp/terraform/internal/command.(*Meta).loadConfig(0xc00037d180, 0x259572d, 0x1, 0xc000351800, 0x1, 0x1, 0x0)
	/home/circleci/project/project/internal/command/meta_config.go:66 +0x133
github.com/hashicorp/terraform/internal/command.(*InitCommand).Run(0xc00037d180, 0xc00009c3d0, 0x0, 0x0, 0xc000598ad0)
	/home/circleci/project/project/internal/command/init.go:189 +0xbac
github.com/mitchellh/cli.(*CLI).Run(0xc0005c28c0, 0xc000887dd0, 0x2596feb, 0x4)
	/home/circleci/go/pkg/mod/github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x41a
main.wrappedMain(0x0)
	/home/circleci/project/project/main.go:384 +0x1132
main.realMain(0x0)
	/home/circleci/project/project/main.go:115 +0x574
main.main()
	/home/circleci/project/project/main.go:58 +0x29

### Expected Behavior

Login to vault

### Actual Behavior

Crash

--- a/internal/configs/config_build.go
+++ b/internal/configs/config_build.go
@@ -23,9 +23,13 @@ func BuildConfig(root *Module, walker ModuleWalker) (*Config, hcl.Diagnostics) {
 	cfg.Root = cfg // Root module is self-referential.
 	cfg.Children, diags = buildChildModules(cfg, walker)
 
-	// Now that the config is built, we can connect the provider names to all
-	// the known types for validation.
-	cfg.resolveProviderTypes()
+	// Skip provider resolution if there are any errors, since the provider
+	// configurations themselves may not be valid.
+	if !diags.HasErrors() {
+		// Now that the config is built, we can connect the provider names to all
+		// the known types for validation.
+		cfg.resolveProviderTypes()
+	}
 
 	diags = append(diags, validateProviderConfigs(nil, cfg, false)...)
 

--- a/internal/configs/testdata/config-diagnostics/invalid-provider/errors
+++ b/internal/configs/testdata/config-diagnostics/invalid-provider/errors
@@ -1,0 +1,1 @@
+main.tf:1,1-20: Invalid provider local name; crash_es is an invalid provider local name

--- a/internal/configs/testdata/config-diagnostics/invalid-provider/main.tf
+++ b/internal/configs/testdata/config-diagnostics/invalid-provider/main.tf
@@ -1,0 +1,3 @@
+module "mod" {
+  source = "./mod"
+}

--- a/internal/configs/testdata/config-diagnostics/invalid-provider/mod/main.tf
+++ b/internal/configs/testdata/config-diagnostics/invalid-provider/mod/main.tf
@@ -1,0 +1,2 @@
+provider "crash_es" {
+}

--- a/internal/configs/testdata/config-diagnostics/invalid-provider/warnings
+++ b/internal/configs/testdata/config-diagnostics/invalid-provider/warnings
@@ -1,0 +1,1 @@
+Empty provider configuration blocks are not required

